File: lr/src/questions/q_0332_reconstruct_itinerary.rs
use std::collections::BTreeMap;

pub fn find_itinerary(tickets: Vec<Vec<String>>) -> Vec<String> {
    let mut graph: BTreeMap<String, Vec<String>> = BTreeMap::new();

    for ticket in tickets {
        graph
            .entry(ticket[0].clone())
            .or_default()
            .push(ticket[1].clone());
    }

    for destinations in graph.values_mut() {
        destinations.sort_by(|a, b| b.cmp(a)); // Sort in reverse order
    }

    let mut route = Vec::new();

    fn dfs(graph: &mut BTreeMap<String, Vec<String>>, airport: &str, route: &mut Vec<String>) {
        while let Some(destination) = graph.get_mut(airport).and_then(|dests| dests.pop()) {
            dfs(graph, &destination, route);
        }
        route.push(airport.to_string());
    }

    dfs(&mut graph, "JFK", &mut route);
    route.reverse();
    route
}

#[test]
fn test_find_itinerary() {
    let tickets = vec![
        vec!["MUC".to_owned(), "LHR".to_owned()],
        vec!["JFK".to_owned(), "MUC".to_owned()],
        vec!["SFO".to_owned(), "SJC".to_owned()],
        vec!["LHR".to_owned(), "SFO".to_owned()],
    ];
    let res = find_itinerary(tickets);
    assert_eq!(
        res,
        vec![
            "JFK".to_owned(),
            "MUC".to_owned(),
            "LHR".to_owned(),
            "SFO".to_owned(),
            "SJC".to_owned()
        ]
    );
}

File: lg/src/questions/q_0332_reconstruct_itinerary.go
package questions

import "slices"

func findItinerary(tickets [][]string) []string {
	graph := make(map[string][]string)

	for _, ticket := range tickets {
		graph[ticket[0]] = append(graph[ticket[0]], ticket[1])
	}

	for _, destinations := range graph {
		slices.Sort(destinations)
	}

	route := make([]string, 0)
	var dfs func(airport string)
	dfs = func(airport string) {
		for len(graph[airport]) > 0 {
			next := graph[airport][0]
			graph[airport] = graph[airport][1:]
			dfs(next)
		}
		route = append([]string{airport}, route...)
	}

	dfs("JFK")
	return route
}
